### PR TITLE
Adding exception handling code and minor code updates to tensor_shape.py

### PR DESCRIPTION
**Pull request by SSaishruthi**
This PR adds error handling code as well as minor updates to the tensor_shape.py

### CODE REVIEW

**Comment by SSaishruthi** [comment]
@gunan This PR has minor code changes to tensor_shape.py.

**Comment by gunan** [comment]
I am not the code owner for this file, please redirect.

**Comment by SSaishruthi** [comment]
@rthadur Can you please assign a reviewer?

**Review by qlzh727 — CHANGES_REQUESTED** [review]
I don't think swallow the TypeError and ValueError is appropriate in this case. The downside here is that the original error contains actionable details about why the error is happening, while the NotImplemented is too generic and doesn't tell user the error detail.

**Comment by SSaishruthi** [comment]
@qlzh727 Thanks for the review. I will add code to print Type and Value error

**Comment by qlzh727** [comment]
What i feel confused is why not just let it throw the original error? What's the benefit to have NotImplemented here comparing to existing error?

**Comment by qlzh727** [comment]
I think I fine with NotImplemented being used at built in binary special methods, eg __eq__, __lt__, etc, so that it will handle those case gracefully, but it shouldn't be used for other case, which explicit error is better. 

Check https://docs.python.org/3/library/constants.html.

**Comment by SSaishruthi** [comment]
@qlzh727 Makes sense. I have updated `is_compatible_with` and `merge_with` methods. Please let me know if this looks good. Thanks in advance.

**Comment by SSaishruthi** [comment]
Hi @qlzh727,
Are updates looking good or need further changes?
TIA

**Comment by qlzh727** [comment]
What I am still bit confused is why this PR is needed in the first place. Can u give a test case that the existing code will fail in an unpleasant way but the current code won't?

**Comment by SSaishruthi** [comment]
Thanks, @qlzh727 for the review. I will work on this and update

**Comment by SSaishruthi** [comment]
Hi @qlzh727 
I am not sure whether to call it an unpleasant way. The current change will directly print the error instead of the traceback. Other methods have similar error handling and that is one of the reasons for the change. 

Current traceback
```
>>> tf.Dimension('-1') == tf.Dimension('1')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/saishruthi.tn@ibm.com/anaconda3/envs/cp/lib/python3.7/site-packages/tensorflow/python/framework/tensor_shape.py", line 190, in __init__
    raise ValueError("Dimension %d must be >= 0" % self._value)
ValueError: Dimension -1 must be >= 0
```

Proposed will only have the error message
```
ValueError: Dimension -1 must be >= 0
```

Please let me know how we can go about this.

File: tensorflow/python/framework/tensor_shape.py [hunk 1]
```diff
@@ -245,7 +245,10 @@ def is_compatible_with(self, other):
     Returns:
       True if this Dimension and `other` are compatible.
     """
-    other = as_dimension(other)
+    try:
+      other = as_dimension(other)
+    except (TypeError, ValueError) as e:
+      return (type(e), '::', e)
```
Comment (qlzh727) [inline tensorflow/python/framework/tensor_shape.py]: I think this is even more confusing and also break the contract of returning a Boolean value. Now its returning a tuple with errors. 

And also, catching the error and rethrow it is kind of anti-pattern in the code. Why do u want to specially handle these errors? Is it just for hiding the stacktrace?

**Review by qlzh727 — CHANGES_REQUESTED** [review]

**Pull request closed without merging**